### ISSUE TITLE
feature request: shared memory concat with new allocation for subsequent operations

### ISSUE THREAD

**Issue by ahundt** [tensorflow/tensorflow#12948]
DenseNet is an effective network design that relies on applying nn layers on recursive concatenations of data along the channel axis. Unfortunately, this has the side effect of quadratic memory growth in TensorFlow as completely new blocks of memory are allocated after each concat operation, resulting in poor performance during all phases of execution.

This is a feature request for a new `allocation='shared'` option for operations such as `tf.concat(allocation='shared')` which works seamlessly with later operations that might modify the data such as BatchNorm, which might also need to share memory. This would make it is possible to utilize the [Memory-Efficient Implementation of DenseNets](http://arxiv.org/abs/1707.06990), a paper which demonstrates that this memory utilization can be dramatically reduced through sharing of allocations. This image from the paper + pytorch implementation illustrates the shared memory approach:

![densenet shared memory](https://camo.githubusercontent.com/d3370ec4935b4bc92b736a122bc226abf48988fd/68747470733a2f2f7261772e6769746875622e636f6d2f67706c656973732f656666696369656e745f64656e73656e65745f7079746f7263682f6d61737465722f696d616765732f666f72776172642e706e67)

- [Pytorch efficient DenseNet implementation](https://github.com/gpleiss/efficient_densenet_pytorch)
- [Keras DenseNet Implementation](https://github.com/titu1994/keras-contrib/blob/ff47da56fbd54cf6cdc2ac2218529fbdadf99296/keras_contrib/applications/densenet.py) with "naive" allocations, works with TensorFlow backend.

This functionality would also be useful for any other application or future network design that employs recursive concatenations. Just in case I didn't find it in my search, perhaps a mechanism already exists that can meet these goals?

**Comment by cy89** [comment]
@zheng-xq, can you please comment?

**Comment by ahundt** [comment]
I'd like to provide some additional motivation for this feature request. 

I'm able to train a Resnet50 based architecture for my problem with a batch size of 32 on the older 12GB Titan X GPU. However, a DenseNet based architecture that should actually be fairly small, with 40 layers and 500k parameters can only fit a batch size of 6, produces some tf memory warnings, and runs 8x slower during training, leading a day of training to take over a week, and smaller batch sizes were even slower. Obviously this isn't a perfect 1:1 comparison since I was using the naive implementation, but the authors indicate they are able to create networks that outperform ResNet in training and evaluation performance on pytorch. To my knowledge, I don't think TF supports an easy to use operations that could be used to re-create this memory efficient implementation.

DenseNet won the 2017 CVPR best paper award, so I think support for capabilities necessary for an efficient DenseNet implementation seem particularly worthwhile for TF. I'd look into implementing it myself but I've already used up my free cycles on a number of other contributions here and in Keras, so I'm sticking to ResNet based solutions at this time. Thanks for your consideration!

**Comment by ybsave** [comment]
Totally agree the necessity of memory shared feature. In addition, to implement memory efficient DenseNet, we also need a feature re-computation function. There are already memory efficient implementations of DenseNet on other platforms; I think Tensorflow, labeled as one of the most flexible platform, falls behind somewhat on this feature. Hope both would be added soon. Thank you.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
this issue is still valid

**Comment by phybrain** [comment]
is project started

**Comment by ysyyork** [comment]
I have the same issue when I'm using densenet from keras-contrib as well as my self implemented unet. Both are using a lot of concat operations. Hope this new feature can be prioritized. Thanks.

**Comment by ahundt** [comment]
It looks like some might use tf.Defun for this, see https://github.com/tensorflow/tensorflow/issues/13160#issuecomment-331914859.
A similar feature for reducing memory with tensorflow might have been implemented outside of tf itself, see https://github.com/tensorflow/tensorflow/issues/13160#issuecomment-357773335.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by YuhuaBillChen** [comment]
I just want to say as for 3D medical images, the 3D DenseNet takes up even more memory dues to the extra dimension of parameters, pushing me to use only 1 or 2 images in a mini-batch, which causes a lot of headache issues during training (Batch normalization won't work in such a small batch, even the batch re-normalization can't help).

**Comment by drpngx** [comment]
Right, I think we are working on something that might help. Stay tuned.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by qianyizhang** [comment]
@drpngx care to elaborate what's the new feature that might help, and what's the progress so far?
I hope it's more than gradient-checkpointing:
https://github.com/openai/gradient-checkpointing

**Comment by drpngx** [comment]
So, we finally pushed the feature. You can implement gradient checkpointing with that as well.

You can take a look at [`recurrent.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/recurrent/python/ops/recurrent.py). It uses the [`inplace_ops`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/inplace_ops.py) that allow to  write to parts of a tensor. Effectively you can manage a large tensor as a random memory access buffer.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ahundt** [comment]
Cool! so we could simply add `alias_inplace_concat` much like `alias_inplace_add`?

```python

def alias_inplace_add(x, i, v):
  """Applies an inplace add on input x at index i with value v. Aliases x.
  If i is None, x and v must be the same shape. Computes
    x += v;
  If i is a scalar, x has a rank 1 higher than v's. Computes
    x[i, :] += v;
  Otherwise, x and v must have the same rank. Computes
    x[i, :] += v;
  Args:
    x: A Tensor.
    i: None, a scalar or a vector.
    v: A Tensor.
  Returns:
    Returns x.
  """
return _inplace_helper(x, i, v, gen_array_ops.inplace_add)
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Yes. I think you can write or customize recurrent to do what you want. If you don't care that the memory is not freed during the computation, then you can pre-allocate to the maximum size and overwrite the parts that you need.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by bhack** [comment]
@drpngx Will this impact in general memory allocation in Shared layers models like https://github.com/tensorflow/tensorflow/issues/16468#issuecomment-385662734?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by drpngx** [comment]
Interesting. So far, this is unconnected, as far as I know. It's an optimization worth considering.

**Comment by joeyearsley** [comment]
@ahundt Did you manage to get it working?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tnbalsam** [comment]
Update: I wrote custom gradient operations for the inplace operations. However, since the inplace update only operates on the 0th dimension, there has to be a transpose on either write or read to the memory block, so training speed doesn't seem to be so hot with that.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by ahundt** [comment]
@joeyearsley sorry I only wanted to document the problem when I noticed it so any fixes (or lack of fixes) can be tracked. I'm not planning to address this myself since it is not critical path for me.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by Ouwen** [comment]
:+1: for this feature, @tnbalsam any progress? If not, happy to try and base off of what you have currently

**Comment by joeyearsley** [comment]
I've made a memory efficient one which mirrors the pytorch version with gradient checkpointing:
https://github.com/joeyearsley/efficient_densenet_tensorflow

Would still be nice to have a shared memory concat though for the future.

**Comment by ahundt** [comment]
Awesome, thanks! I'll close this now.

**Comment by Ouwen** [comment]
@joeyearsley this is great, thanks for the reference code
:+1: for the future memory concat feature

**Comment by Ouwen** [comment]
@ahundt it may be worth re-opening this as having concat shared memory is still a huge memory advantage even with gradient checkpointing.

**Comment by Ouwen** [comment]
@drpngx I was wondering if you can provide some guidance here

The [`inplace_ops`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/inplace_ops.py) are great that they allow you to do **writes** to an already memory allocated tensor, but it would be nice to have inplace **reads**.

* Perhaps we could modify `tf.slice` to have an option to not allocate new memory for an output tensor.

* I was also wondering if `tf.TensorArray` could be utilized here instead of the `inplace_ops` where we just `concat` as the `tf.TensorArray` grows. However, it seems like this would be inefficient on GPUs (#19733)

**Comment by drpngx** [comment]
The writes work like `assign`. After they are executed, the tensor is modified. If you want to read memory, as you point out, you can use `tf.slice` or `tf.Tensor.__getitem__` (the `[]` operator).

**Comment by Ouwen** [comment]
Ah I see so `tf.slice` will not allocate new memory?

**Comment by byronyi** [comment]
@ahundt @Ouwen If you are still interested, the [ScopeAllcator](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/common_runtime/scoped_allocator.h) introduced lately could be of help for implementing this feature.
